### PR TITLE
[WIP] Try to replace all gulp with webpack

### PR DESCRIPTION
**Pull request by bryanbecker**
Ref: https://github.com/screepers/screeps-typescript-starter/issues/44

I consider this PR complete

**Notes**:

Testing is not implemented whatsoever.  I'm leaving that task to someone else who wants to test their screeps code (ha).  I imagine it's extremely straightforward to set up, though.

The webpack version of the source map / git interaction can not automatically pick up the repository path, as it did before.  The path now needs to be manually set in the config.  There are two large "TODO"s in the Readme where documentation on how to accomplish this needs to be added.  Since I don't use this feature, I'm not sure how it should be set.

The `screeps-webpack-plugin` is set to install from my fork, pending this PR: https://github.com/langri-sha/screeps-webpack-plugin/pull/28

**Features:**
- version upped to 2.0.0 (from 0.0, hah), since many config changes are breaking
- updated readme
- typescript update to latest ~(2.4.0)~ 2.3.2
- ts-loader replaced with awesome-typescript-loader for async compile and type-checking
- "snippets" dir, mostly to serve as an example of excluding folders
- typescript `baseUrl` option set, so that modules can be included without using relative paths (relative paths still work, though)
- uploading to private servers works
- build toggles work that are replaced at build time (defined in config)
- git integration added (can create toggles / change build target based on git info)
- add screeps profiler
- removed outdated example of toggling pathfinder
- add new example of toggling profiler

### CODE REVIEW

File: config/credentials.example.json [hunk 1]
```diff
@@ -0,0 +1,8 @@
+{
+  "email": 'EMAIL',
```
Comment (resir014) [inline config/credentials.example.json]: replace the single quotes on this file with double quotes, as per the [JSON specifications](http://www.json.org/).

File: package.json [hunk 2]
```diff
@@ -1,78 +1,35 @@
...
-    "start": "gulp watch",
-    "prod": "gulp watch --target prod",
-    "test": "mocha",
-    "test:coverage": "nyc mocha",
-    "tslint": "tslint \"src/**/*.ts\" \"test/**/*.ts\"",
-    "typings": "typings",
-    "jshint-gulp": "jshint gulpfile.js libs/gulp-screeps-upload.js"
+    "deploy": "webpack --colors",
+    "start": "NODE_ENV=dev webpack --colors --watch",
+    "test": "echo \"Error: no test specified\" && exit 1",
```
Comment (resir014) [inline package.json]: This should work fine with `mocha`. I'll test it on my personal repo just to make sure.

Not sure about coverage though.

**Review by resir014 — CHANGES_REQUESTED** [review]
Awesome work! I've put some minor style guide suggestions, but other than that, it's good to go.

**Comment by resir014** [comment]
I've also done some more extensive tests on, well, tests, and I've managed to run the tests fine with minimal tweaks.

You'll need to add the following packages:

```sh
$ npm install --save-dev @types/mocha @types/chai mocha chai ts-node
```

**One little caveat:** Setting `strict: true` on your `tsconfig.json` (basically, using strict mode) will cause the tests to break due to `noImplicitAny` errors on `test/mock/game.ts` (the whole file mocks the entire game's constants so some TS rules are ignored). Can we actually suppress these errors for this one file only, or do we need to find a better way to mock game states?

**Comment by bryanbecker** [comment]
What is causing the errors... mocha or tslint?  We can add files or directories to be excluded by tslint (or any module, really), in the config (see [this line](https://github.com/screepers/screeps-typescript-starter/pull/45/files#diff-302de8d679dbd724686b4a0b44b48f0cR109))

I couldn't get the tests to run via a simple `./node_modules/mocha/bin/mocha` command, but that may be because it expects things to be compiled (complained about `__REVISION__` is not set)

Re integration, I found two modules:

[mocha webpack](https://github.com/zinserjan/mocha-webpack) which seems more full features.  It's run with the `mocha-webpack` command, and has a watch mode.  Not sure if it integrates well into our existing webpack config, though (i.e. we may be face with with "pick one: mocha-webpack watch or webpack watch"

[mocha loader](https://github.com/webpack-contrib/mocha-loader) this seems the safer bet to me.  It integrates into the existing webpack config, and is run as part of the build.

Question: is [sinon](https://github.com/sinonjs/sinon/) still needed in the package?  Sinon, jshint, intanbul & nyc seem to be the major packages that haven't made it into "2.0", yet

Last, is testing "opt-in" or "opt-out".  There's a few ways to put it into the existing config.

1. Add testing into `config.default.js`, with README telling how to remove.  This will enforce testing in all environments unless specifically removed in an environment config
2. Add to `config.dev.js`.  This will have the same affect for people just starting, since `dev` is  the only environment
3. Make new environment file
4. "Hack" it with env vars like so:

```js
  // regular config to run testing with any file matching /test.js$/ regex (except in node_modules)
    rules: [{
      test: /test.js$/,
      use: 'mocha-loader',
      exclude: /node_modules/,
    }]

  // HACK with something like:
if ("[testing] === true") {
  Config.module.rules.push(\* the above rule object *\);
}
```

Setting up the config for option 4 has a lot of interesting tradeoffs.

For one, it will encourage a minor config refactor.  While the config currently is just a really large object with values filled in.  We could refactor to something like:

```js
//... snip
module: = getModules();
// .. snip

function getModules() {
  // returns the modules object
}
```

The tradeoff is that it might accidentally balloon into a back-from-the-dead gulpfile monstrosity

**Comment by bryanbecker** [comment]
Hold off on the merge.  I made another local branch with a config that I think is much more maintainable and organized.

Still need to make sure a few things work correctly, though

**Comment by resir014** [comment]
@bryanbecker 

> What is causing the errors... mocha or tslint?

I think it was `mocha`, since the test is written on TS as well we had to use the `ts-node` runtime when passing the code to `mocha`, which gives the compile errors

**Review by resir014 — APPROVED** [review]

**Comment by resir014** [comment]
Anyways, I approved the changes but I'll wait for your end of the work, then I'll give it a go in my codebase again.

**Comment by bryanbecker** [comment]
OK, everything is in typescript now... still need to update the README to match, but config is much easier to programmatically alter, now

Re:  Mocha:  I think some additional tool is needed (Karma, maybe?), since it needs to compile the TS before running the tests.

[Mocha Webpack](https://github.com/zinserjan/mocha-webpack)  I think can handle it, too... The issue is that it needs to know where to import the files/modules from

**Comment by resir014** [comment]
Yeah, you're right. /:

This might take a while to figure out, so do you want me to merge this and look into re-integrating the testing framework on a separate PR?

**Comment by bryanbecker** [comment]
Sure, let me update the README, then lets merge and let some people try it out

**Comment by bryanbecker** [comment]
Done

**Comment by resir014** [comment]
Alright, thanks! Merging now.

**Pull request merged**